bc_obps/compliance/service/penalty_calculation_service.py

```diff
@@ -327,4 +444,176 @@ def create_penalty(cls, obligation: ComplianceObligation) -> CompliancePenalty:
+        if not interest_rates.exists():
+            interest_rates = ElicensingInterestRate.objects.filter(is_current_rate=True)
+
+        if not interest_rates.exists():
+            raise ValueError("No interest rates configured in the system")
+
+        def get_rate_for_date(date_to_check: date) -> Decimal:
+            """Helper function to get the interest rate for a specific date"""
+            for rate in interest_rates:
+                if rate.start_date <= date_to_check <= rate.end_date:
+                    return rate.interest_rate
+
+            return interest_rates[0].interest_rate
+
+        # Calculate penalty for each monthly accrual date
+        for accrual_date in monthly_accrual_dates:
+            annual_rate = get_rate_for_date(accrual_date)
+            monthly_rate = annual_rate / 12
+            payments = cls.sum_payments_before_date(invoice, accrual_date)
+            adjustments = cls.sum_adjustments_before_date(invoice, accrual_date)
+            penalty_amount = (base - payments + adjustments) * monthly_rate
+            monthly_compounding = (accumulated_penalty + accumulated_compounding) * monthly_rate
+            accumulated_penalty += penalty_amount
+            accumulated_compounding += monthly_compounding
+            total_penalty = accumulated_penalty + accumulated_compounding
+
+            if persist_penalty_data:
+                CompliancePenaltyAccrual.objects.create(
+                    compliance_penalty=compliance_penalty_record,
+                    date=accrual_date,
+                    interest_rate=monthly_rate.quantize(Decimal('0.0001'), rounding=ROUND_HALF_UP),
+                    daily_penalty=penalty_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+                    daily_compounded=monthly_compounding.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+                    accumulated_penalty=accumulated_penalty.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+                    accumulated_compounded=accumulated_compounding.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+                )
+
+        if persist_penalty_data:
+            compliance_penalty_record.penalty_amount = total_penalty.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
+            compliance_penalty_record.save()
+
+        result = {
+            "penalty_status": obligation.penalty_status,
+            "penalty_type": CompliancePenalty.PenaltyType.LATE_SUBMISSION,
+            "penalty_charge_rate": (get_rate_for_date(final_accrual_date) * 100) / 12,
+            "days_late": days_late,
+            "accumulated_penalty": accumulated_penalty.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+            "accumulated_compounding": accumulated_compounding.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+            "total_penalty": total_penalty.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP),
+        }
+
+        return result
```

Comment:
Are we using this return value anywhere, or planning to use it in the future?
If not, we can simply return the newly created penalty when `persist_penalty_data` is True; otherwise, return None. That way, in the `create_late_submission_penalty` function, we don’t need to query like this:

```python
CompliancePenalty.objects.filter(
            compliance_obligation=obligation,
            penalty_type=CompliancePenalty.PenaltyType.LATE_SUBMISSION,
        ).latest('id')
```